core/azure-core/src/main/java/com/azure/core/polling/Poller.java

```diff
@@ -0,0 +1,275 @@
+     * @param cancelOperation cancel operation
+     */
+    public Poller(PollerOptions pollerOptions, Function<PollResponse<T>, PollResponse<T>> pollOperation, Consumer<Poller> cancelOperation) {
+        this(pollerOptions, pollOperation);
+        this.cancelOperation = cancelOperation;
+    }
+
+    /**
+     * This will call cancelOperation function if provided. Once cancelOperation is triggered successfully, it can not be called  again.
+     * It will not call cancelOperation if operation status is not started/Cancelled/Failed/successfully completed.
+     *
+     * @see com.azure.core.polling.PollResponse.OperationStatus
+     * @throws UnsupportedOperationException when cancel operation is not provided.
+     */
+    public void cancelOperation() throws UnsupportedOperationException {
+        if (cancelOperation == null) {
+            throw new UnsupportedOperationException("Cancel operation is not supported on this service/resource.");
+        }
+
+        // We can not cancel an operation if it was never started
+        // or it is in its terminal state.
+        // Check cancelInitiated: to protect against multiple time call to cancel operation
+        if (cancelInitiated || (pollResponse != null && pollResponse.getStatus() != PollResponse.OperationStatus.IN_PROGRESS)) {
+            return;
+        }
+        //Time to call cancel
+        cancelOperation.accept(this);
+        cancelInitiated = true;
+    }
+    /**
+     * Enable user to subscribe and receive all the responses.
+     * The user will start receiving PollResponse When client subscribe to this Flux.
+     * The poller still have its own default polling in action unless, user has  turned off
+     * auto polling.
+     *
+     * @return poll response as Flux that can be subscribed.
+     */
+    public Flux<PollResponse<T>> getObserver() {
+        return sendPollRequestWithDelay()
+            .flux()
+            .repeat()
+            .takeUntil(pollResponse -> pollResponse.isDone());
+    }
+
+    /*
+     * We will maintain one instance of backgroundFluxHandle per poller.
+     * This handle will be set to null when autopolling is disabled.
+     */
+    private Flux<PollResponse<T>> createBackgroundFlux() {
+        if (backgroundFluxHandle == null) {
+            backgroundFluxHandle = sendPollRequestWithDelayBackground()
+                .flux()
+                .repeat()
+                .takeUntil(pollResponse -> pollResponse.isDone());
+        }
+        return backgroundFluxHandle;
+    }
+
+    /**
+     * Calls poll operation once in sync if Pool operation is not completed.
+     * @return a Mono of {@link PollResponse}
+     */
+    public Mono<PollResponse<T>> poll() {
+
+        if (!isTerminalState()) {
+            updatePollOperationSync();
+        }
+        return Mono.just(pollResponse);
```

Comment:
Need to reconcile if you want things to be sync or async. At the moment, as @jianghaolu points out, there is a bit of a mismatch going on. The poll operation isn't really async, but it acts as if it is.

Reply:
This method will return the old `pollResponse` value, as the `updatePollOperationAsync()` method call is operating asynchronously, and won't necessarily have updated the `pollResponse` field. Same issue appears to exist in the `sendPollRequestWithDelay` method below.

Reply:
resolved